### PR TITLE
refactor(api): move heater shaker hardware interaction to core

### PR DESCRIPTION
**Pull request by mcous**
## Overview

Heater-shaker portion of RCORE-103

## Changelog

- Move heater shaker context hardware interactions to `LegacyHeaterShakerCore`
- Remove unused `HeaterShaker` hardware interfaces

## Review requests

Full module testing. Jupyter Notebook used for testing: [PAPIv2 Module Regression Testing.ipynb.zip](https://github.com/Opentrons/opentrons/files/9747790/PAPIv2.Module.Regression.Testing.ipynb.zip)

## Risk assessment

Anticipating this will be low. Heater-Shaker is our newest modules and its APIs have incurred the least tech-debt of any of our modules

### CODE REVIEW

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11480** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11480" target="_blank">Graphite</a> 
    * **PR #11496** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11496" target="_blank">Graphite</a> 
      * **PR #11497** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11497" target="_blank">Graphite</a> 
        * **PR #11504** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11504" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11504?utm_source=stack-comment).

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    decoy: Decoy,
+    mock_sync_module_hardware: SyncHeaterShakerHardware,
+    mock_geometry: HeaterShakerGeometry,
+    mock_protocol_core: ProtocolContextImplementation,
+    subject: LegacyHeaterShakerCore,
+) -> None:
+    """It should open the labware latch with the hardware."""
+    mock_hardware = decoy.mock(cls=SyncHardwareAPI)
+    decoy.when(subject.get_speed_status()).then_return(SpeedStatus.IDLE)
+    decoy.when(mock_geometry.is_pipette_blocking_latch_movement(None)).then_return(True)
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: We should not be checking for `None` here, we should be stubbing `mock_protocol_core.get_last_location()` and ensure the return value is passed into `mock_geometry.is_pipette_blocking_latch_movement`

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    decoy.when(mock_protocol_core.get_hardware()).then_return(mock_hardware)
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Move this stubbing to the `mock_protocol_core` fixture. Longer term, I'd want to make `ProtocolCore.get_hardware()` a property instead of a get method

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+def test_open_labware_latch(
+    decoy: Decoy,
+    mock_sync_module_hardware: SyncHeaterShakerHardware,
+    mock_geometry: HeaterShakerGeometry,
+    mock_protocol_core: ProtocolContextImplementation,
+    subject: LegacyHeaterShakerCore,
+) -> None:
+    """It should open the labware latch with the hardware."""
+    mock_hardware = decoy.mock(cls=SyncHardwareAPI)
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Create `mock_hardware` fixture

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    decoy.when(subject.get_target_temperature()).then_return(42.0)
+
+    subject.wait_for_target_temperature()
+
+    decoy.verify(mock_sync_module_hardware.await_temperature(), times=1)
+
+
+def test_wait_for_temperature_no_target(
+    decoy: Decoy,
+    mock_sync_module_hardware: SyncHeaterShakerHardware,
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Unused

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        HeaterShakerLabwareLatchStatus.UNKNOWN,
+        HeaterShakerLabwareLatchStatus.CLOSING,
+        HeaterShakerLabwareLatchStatus.IDLE_OPEN,
+        HeaterShakerLabwareLatchStatus.OPENING,
+    ],
+)
+def test_set_and_wait_for_shake_speed_raises(
+    decoy: Decoy,
+    latch_status: HeaterShakerLabwareLatchStatus,
+    mock_sync_module_hardware: SyncHeaterShakerHardware,
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Unused

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        SpeedStatus.DECELERATING,
+        SpeedStatus.ACCELERATING,
+        SpeedStatus.HOLDING,
+        SpeedStatus.ERROR,
+    ],
+)
+def test_open_labware_latch_raises(
+    decoy: Decoy,
+    speed_status: SpeedStatus,
+    mock_sync_module_hardware: SyncHeaterShakerHardware,
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Unused

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    decoy.when(mock_geometry.is_pipette_blocking_latch_movement(None)).then_return(True)
+    decoy.when(mock_protocol_core.get_hardware()).then_return(mock_hardware)
+
+    subject.open_labware_latch()
+
+    decoy.verify(
+        mock_hardware.home(axes=[Axis.Z, Axis.A]),
+        mock_protocol_core.set_last_location(None),
+        mock_sync_module_hardware.open_labware_latch(),
+        times=1,
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py:339]: Remove `times` argument for call sequences

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+    mock_geometry: HeaterShakerGeometry,
+    mock_protocol_core: ProtocolContextImplementation,
+    subject: LegacyHeaterShakerCore,
+) -> None:
+    """It should set and wait for the target speed with the hardware."""
+    mock_hardware = decoy.mock(cls=SyncHardwareAPI)
+    decoy.when(subject.get_labware_latch_status()).then_return(
+        HeaterShakerLabwareLatchStatus.IDLE_CLOSED
+    )
+    decoy.when(mock_geometry.is_pipette_blocking_shake_movement(None)).then_return(True)
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Same comment: do not use `None` here, stub out the dependency that feeds this argument.

Also, do we need corresponding tests for `False`, or is that covered elsewhere?

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+def test_set_and_wait_for_shake_speed(
+    decoy: Decoy,
+    mock_sync_module_hardware: SyncHeaterShakerHardware,
+    mock_geometry: HeaterShakerGeometry,
+    mock_protocol_core: ProtocolContextImplementation,
+    subject: LegacyHeaterShakerCore,
+) -> None:
+    """It should set and wait for the target speed with the hardware."""
+    mock_hardware = decoy.mock(cls=SyncHardwareAPI)
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Same comment: move this to a common fixture

File: api/src/opentrons/hardware_control/modules/heater_shaker.py [hunk 10]
```diff
@@ -309,22 +279,26 @@ async def start_set_temperature(self, celsius: float) -> None:
         await self._driver.set_temperature(celsius)
         await self.wait_next_poll()
 
-    async def await_temperature(self, awaiting_temperature: float) -> None:
+    async def await_temperature(
+        self, awaiting_temperature: Optional[float] = None
+    ) -> None:
         """Await temperature in degree Celsius.
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/heater_shaker.py]:
```suggestion
        """Await temperature in degrees Celsius.
```

File: api/src/opentrons/protocol_api/core/protocol_api/legacy_module_core.py [hunk 11]
```diff
@@ -1,42 +1,63 @@
 """Legacy Protocol API module implementation logic."""
+from __future__ import annotations
+
 import logging
-from typing import Optional, cast
+from typing import Optional, cast, TYPE_CHECKING
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/legacy_module_core.py]:
```suggestion
from typing import TYPE_CHECKING, Optional, cast
```

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 12]
```diff
@@ -785,66 +786,67 @@ def labware_latch_status(self) -> str:
...
     def wait_for_temperature(self) -> None:
         """Delays protocol execution until the Heater-Shaker has reached its target
         temperature. Returns an error if no target temperature was previously set.
         """
-        if self.target_temperature is None:
-            raise NoTargetTemperatureSetError(
-                f"Heater-Shaker Module {self} does not have a target temperature set."
-            )
-        self._module.await_temperature(awaiting_temperature=self.target_temperature)
+        # TODO should we pass the target temperature to the core (and then to await temperature?)
```
Comment (mcous) [inline api/src/opentrons/protocol_api/module_contexts.py]: I don't think so, this is now properly handled in the core

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 13]
```diff
@@ -859,33 +861,40 @@ def open_labware_latch(self) -> None:
...
+        self._core.close_labware_latch()
 
     @requires_version(2, 13)
     @publish(command=cmds.heater_shaker_deactivate_shaker)
     def deactivate_shaker(self) -> None:
-        """Stops shaking. Decelerating to 0 rpm typically only takes a few seconds."""
-        self._module.deactivate_shaker()
+        """Stops shaking.
+
+        Decelerating to 0 rpm typically only takes a few seconds.
```
Comment (mcous) [inline api/src/opentrons/protocol_api/module_contexts.py]: Is it `rpm` or `RPM`? I don't remember where we landed

**Review by mcous — COMMENTED** [review]
Final cleanup items, I think

File: api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py (hunk 8 above)
Comment (jbleon95) [reply api/tests/opentrons/protocol_api/core/legacy/test_heater_shaker_core.py]: Made an additional test for this and `open_latch` if pipette is not blocking

File: api/src/opentrons/protocol_api/module_contexts.py (hunk 13 above)
Comment (jbleon95) [reply api/src/opentrons/protocol_api/module_contexts.py]: every other non-argument related doc string reference to RPM uses all caps, so changed it here as well

**Review by mcous — COMMENTED** [review]
This LGTM. Let's undraft and move into temp deck, mag deck, and H/S testing. We can merge #11496, #11497, and this one in sequence once testing has been completed on this branch

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

File: api/src/opentrons/hardware_control/modules/heater_shaker.py [hunk 14]
```diff
@@ -309,22 +279,26 @@ async def start_set_temperature(self, celsius: float) -> None:
         await self._driver.set_temperature(celsius)
         await self.wait_next_poll()
 
-    async def await_temperature(self, awaiting_temperature: float) -> None:
-        """Await temperature in degree Celsius.
+    async def await_temperature(
+        self, awaiting_temperature: Optional[float] = None
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/heater_shaker.py]: Let's revert this change in favor of a TODO, since it's a behavioral change and runs some risk, especially given error statuses

**Comment by mcous** [comment]
Tested this on an OT-2 with a couple H/S units, and everything behaved as expected. We encountered RET-1259, but confirmed that it's an existing bug with the Hardware API's module polling and not introduced or changed with this PR

**Comment by mcous** [comment]
**Graphite Merge Job**

Current status: ✅ Merged

This pull request was successfully merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11504?utm_source=stack-comment).


Job Reference: MV0Dt42KFXlpVMpKUJh2
